irc/user.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+import { Conn } from "deno";
+
+const utf8Decoder = new TextDecoder();
+const utf8Encoder = new TextEncoder();
+const DEFAULT_BUFFER_SIZE = 1024;
+
+type UserMode =
+  | "+i"
+  | "+o"
+  | "+O"
+  // | "+r"
+  | "+w";
+
+export class User {
+  private _conn: Conn;
+  private _buf: Uint8Array;
+
+  // TODO: buffer full IRC messages
+  private _messageBuf: string = "";
+
+  private _userModes: Set<UserMode> = new Set();
+  public nickname = "";
+  public username = "";
+  public fullname = "";
+  public isRegistered = false;
+  public id = "";
+
+  constructor(conn: Conn) {
+    this._conn = conn;
+    this._buf = new Uint8Array(DEFAULT_BUFFER_SIZE);
+  }
+
+  async *readMessages(): AsyncIterableIterator<string> {
+    while (true) {
+      const rr = await this._conn.read(this._buf);
+      if (rr.eof) {
+        return;
+      }
+
+      const plaintext = utf8Decoder.decode(this._buf.buffer.slice(0, rr.nread));
+      // TODO(fancyplants) figure out how to split incoming multiple messages
```

Comment:
This is the only major design flaw I see in your patch.

Here you should either use `BufReader` or `TextProtoReader` - ideally the latter if it works for the IRC proto (I think it will) Using this library you can safely read a line:

https://github.com/denoland/deno_std/blob/41a2d218264b2f9217bf793893aff5dadd2c4ca9/net/textproto_test.ts#L15-L28